### PR TITLE
Add lumen.ai

### PR DESCRIPTION
**Pull request by philippjfr**

### CODE REVIEW

**Review by ahuang11 — COMMENTED** [review]
Looks awesome! Thanks for refactoring the code; very insightful to see how you changed it.

**Comment by ahuang11** [comment]
I just tried this out; here are the necessary additional deps to run this:
```
 1640  pip install pydantic
 1642  pip install instructor
 1644  pip install huggingface_hub
 1646  CMAKE_ARGS="-DLLAMA_METAL=on" pip install -U llama-cpp-python --no-cache-dir\n
 1648  pip install -U panel
```

However, I couldn't get it working:
```
        while (unmet_dependencies:=tuple(r for r in subagent.requires if r not in memory)):
            subagents = [
                agent for agent in self.agents if any(ur in agent.provides for ur in unmet_dependencies)
            ]
```

I added this line:
```

            if not subagents:
                print(f"Failing to find any subagents to handle the remaining requirements: {unmet_dependencies}.")
                continue
```

The memory seems to not have shared this:
`Failing to find any subagents to handle the remaining requirements: ('current_source',).`

**Comment by philippjfr** [comment]
Not sure what precisely was causing that in older Panel versions but the memory works correctly with Panel master.

**Comment by philippjfr** [comment]
Made a bunch of updates but one thing I haven't solved is how to deal with the in-memory data source correctly. Since it can't be serialized properly (at least for now), the LumenBaseAgent can't currently render it correctly. I had hacked that for my demo but it needs a proper fix.

File: lumen/ai/agents.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+
+    provides = param.List(default=['current_plot'], readonly=True)
+
+    def _view_prompt(self, model: BaseModel, view: hvPlotUIView, table: str, schema: dict) -> str:
+        doc = view.__doc__.split('\n\n')[0]
+        prompt = f'{doc}'
+        prompt += f'\nThe arguments must conform to the following schema:\n\n```{model.model_json_schema()}```'
+        prompt += f'\n\nThe data follows the following JSON schema:\n\n```{str(schema)}```'
+        if 'current_view' in memory:
+            prompt += f"The previous view specification was: {memory['current_view']}"
```
[the remark this answers is not in the document]
Comment (philippjfr) [reply lumen/ai/agents.py:349]: Could maybe move onto the `LumenBaseAgent` and be renamed `_component_prompt`.

File: lumen/ai/agents.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        for chunk in self.llm.stream(messages, system=system_prompt, response_model=self.response_model):
+            if not chunk:
+                continue
+            message = self.interface.stream(
+                chunk, user=self.user, message=message, replace=True
+            )
+        return message
+
+
+class SourceAgent(Agent):
```
Comment (philippjfr) [inline lumen/ai/agents.py:84]: Needs to be massively refactored, also maybe this shouldn't be an agent at all or it should probably ask the user what they want to do, upload or provide a URL.

File: lumen/ai/agents.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+    Should only be used if the user explicitly requests adding a source
+    or no source is in memory.
+    """
+
+    requires = param.List(default=[], readonly=True)
+
+    provides = param.List(default=['current_source'], readonly=True)
+
+    on_init = param.Boolean(default=True)
```
Comment (philippjfr) [inline lumen/ai/agents.py:96]:
```suggestion
```

File: lumen/ai/agents.py [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+            a.parentNode.removeChild(a);  //afterwards we remove the element again
+            """,
+        )
+        icons = pn.Row(copy_icon, download_icon)
+        code_col = pn.Column(code_editor, icons, sizing_mode="stretch_both")
+        placeholder = pn.Column()
+        tabs = pn.Tabs(
+            ("Code", code_col),
+            ("Output", placeholder),
+        )
```
Comment (ahuang11) [inline lumen/ai/agents.py:95]: Setting `dynamic=True` here results in 
```
Traceback (most recent call last):
  File "/Users/ahuang/repos/panel/panel/viewable.py", line 604, in _preprocess
    hook(self, root, changed, old_models)
TypeError: link_param_method() takes 2 positional arguments but 4 were given

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ahuang/repos/panel/panel/io/document.py", line 357, in unlocked
    raise e
  File "/Users/ahuang/repos/panel/panel/io/document.py", line 316, in unlocked
    yield
  File "/Users/ahuang/repos/panel/panel/reactive.py", line 301, in _apply_update
    self._update_model(events, msg, root, model, doc, comm)
  File "/Users/ahuang/repos/panel/panel/layout/base.py", line 134, in _update_model
    state._views[ref][0]._preprocess(root, self, old_children)
  File "/Users/ahuang/repos/panel/panel/viewable.py", line 606, in _preprocess
    hook(self, root)
  File "/Users/ahuang/repos/panel/panel/param.py", line 1336, in link_param_method
    methods = root_view.select(lambda p: isinstance(p, ParamMethod) and p.loading_indicator)
  File "/Users/ahuang/repos/panel/panel/chat/feed.py", line 850, in select
    return selected + self._card.select(selector)
  File "/Users/ahuang/repos/panel/panel/layout/card.py", line 92, in select
    return self._header_layout.select(selector) + super().select(selector)
  File "/Users/ahuang/repos/panel/panel/layout/base.py", line 343, in select
    objects += obj.select(selector)
  File "/Users/ahuang/repos/panel/panel/layout/base.py", line 343, in select
    objects += obj.select(selector)
  File "/Users/ahuang/repos/panel/panel/chat/message.py", line 722, in select
    return super().select(selector) + self._composite.select(selector)
  File "/Users/ahuang/repos/panel/panel/layout/base.py", line 343, in select
    objects += obj.select(selector)
  File "/Users/ahuang/repos/panel/panel/layout/base.py", line 343, in select
    objects += obj.select(selector)
  File "/Users/ahuang/repos/panel/panel/layout/base.py", line 343, in select
    objects += obj.select(selector)
  [Previous line repeated 2 more times]
AttributeError: 'function' object has no attribute 'select'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/tornado/ioloop.py", line 738, in _run_callback
    ret = callback()
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/tornado/ioloop.py", line 762, in _discard_future_result
    future.result()
  File "/Users/ahuang/repos/panel/panel/io/server.py", line 184, in wrapper
    state._handle_exception(e)
  File "/Users/ahuang/repos/panel/panel/io/state.py", line 445, in _handle_exception
    raise exception
  File "/Users/ahuang/repos/panel/panel/io/server.py", line 182, in wrapper
    return await func(*args, **kw)
  File "/Users/ahuang/repos/panel/panel/chat/feed.py", line 538, in _prepare_response
    await self._cleanup_response()
  File "/Users/ahuang/repos/panel/panel/chat/interface.py", line 602, in _cleanup_response
    await super()._cleanup_response()
  File "/Users/ahuang/repos/panel/panel/chat/feed.py", line 544, in _cleanup_response
    with param.parameterized.batch_call_watchers(self):
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/contextlib.py", line 142, in __exit__
    next(self.gen)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/param/parameterized.py", line 307, in batch_call_watchers
    parameterized.param._batch_call_watchers()
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/param/parameterized.py", line 2506, in _batch_call_watchers
    self_._execute_watcher(watcher, events)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/param/parameterized.py", line 2468, in _execute_watcher
    watcher.fn(*args, **kwargs)
  File "/Users/ahuang/repos/panel/panel/reactive.py", line 675, in link_cb
    setattr(target, links[event.name], event.new)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/param/parameterized.py", line 525, in _f
    instance_param.__set__(obj, val)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/param/parameterized.py", line 527, in _f
    return f(self, obj, val)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/param/parameterized.py", line 1545, in __set__
    obj.param._call_watcher(watcher, event)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/param/parameterized.py", line 2486, in _call_watcher
    self_._execute_watcher(watcher, (event,))
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/param/parameterized.py", line 2468, in _execute_watcher
    watcher.fn(*args, **kwargs)
  File "/Users/ahuang/repos/panel/panel/reactive.py", line 373, in _param_change
    self._apply_update(named_events, properties, model, ref)
  File "/Users/ahuang/repos/panel/panel/reactive.py", line 300, in _apply_update
    with unlocked():
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/ahuang/repos/panel/panel/io/document.py", line 377, in unlocked
    curdoc.unhold()
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/bokeh/document/document.py", line 776, in unhold
    self.callbacks.unhold()
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 431, in unhold
    self.trigger_on_change(event)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 413, in trigger_on_change
    invoke_with_curdoc(doc, invoke_callbacks)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 443, in invoke_with_curdoc
    return f()
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 412, in invoke_callbacks
    cb(event)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 276, in <lambda>
    self._change_callbacks[receiver] = lambda event: event.dispatch(receiver)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/bokeh/document/events.py", line 353, in dispatch
    super().dispatch(receiver)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/bokeh/document/events.py", line 219, in dispatch
    cast(DocumentPatchedMixin, receiver)._document_patched(self)
  File "/Users/ahuang/miniconda3/envs/lumen/lib/python3.10/site-packages/bokeh/server/session.py", line 244, in _document_patched
    raise RuntimeError("_pending_writes should be non-None when we have a document lock, and we should have the lock when the document changes")
RuntimeError: _pending_writes should be non-None when we have a document lock, and we should have the lock when the document changes
```

File: lumen/ai/agents.py [hunk 5]
```diff
@@ -332,12 +332,16 @@ def format_float(num):
             else:
                 return f"{num:.1e}"  # Exponential notation with two decimals
 
-        if len(df) < 100:
+        length = len(df)
+        if length < 100:
```
Comment (ahuang11) [inline lumen/ai/agents.py:336]: Should df.size be used here? I believe len() only checks rows while size checks cols and rows.

File: lumen/ai/agents.py [hunk 6]
```diff
@@ -391,9 +393,16 @@ def format_float(num):
...
                 df_head_dict[col] = df_head_dict[col].tolist()
 
-        data = {"stats": df_describe_dict, "head": df_head_dict}
-        data_string = str(data)
-        return data_string
+        data = {
+            "summary": {
+                "total_length": length,
+                "is_summarized": is_summarized,
+                "dtypes": {col: str(dtype) for col, dtype in df.dtypes.to_dict().items()}
```
Comment (ahuang11) [inline lumen/ai/agents.py:400]: For dtypes, I did:
```python
        # df_dtypes_dict = {
        #     col: str(type(df[col].iloc[:1].astype("object").iloc[0])).split("'")[1]
        #     for col in df.columns
        # }
```
To reduce the number of tokens and also make it into native python dtypes if I recall correctly (e.g. dtype('int64') -> int)

File: lumen/ai/agents.py (hunk 5 above)
Comment (philippjfr) [reply lumen/ai/agents.py:336]: Feel free to fix.

File: lumen/ai/agents.py (hunk 6 above)
Comment (philippjfr) [reply lumen/ai/agents.py:400]: Please fix.

**Pull request merged**